### ISSUE TITLE
No MP_JOIN segments seen following ADD_ADDR

### ISSUE THREAD

**Issue by kpeletidis** [multipath-tcp/mptcp#251]
I am trying to understand why I see additional subflows being formed only when traffic flows in a particular direction.

This experiment involved two hosts, A and B.

Host A has an ethernet interface, eth0. Three UDP (fou) tunnels, ipudp0010[1-3], are established on top of eth0. The remote addresses are those of the interfaces of host B.

Host B has three interfaces, eth[0-2]. A fou tunnel is established on top of each ethernet interface. Each one has the address of host A as its remote address.

When host B sends traffic to the iperf server on host A it can be seen that traffic flows through all available tunnels.

**When host A sends traffic to the iperf server on host B traffic flows through only one interface.**

According to the packet capture host A receives ADD_ADDR segments with the remaining IP addresses of host B but it doesn't send any SYN+MP_JOIN segments.

I have enabled mptcp debug messages but didn't see any errors (see dmesg output below). My understanding is that host A should start sending SYN+MP_JOIN segments once it knows the peer IP addresses. Any help would be appreciated.

Host A
------

Linux-4.14.24

$ cat /proc/net/mptcp_fullmesh 
Index, Address-ID, Backup, IP-address, if-idx
IPv4, next v4-index: 2
1, 2, 0, 172.31.39.81 2
IPv6, next v6-index: 0

$ ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc htb state UP mode DEFAULT group default qlen 1000
    link/ether 02:e7:98:5b:9f:32 brd ff:ff:ff:ff:ff:ff
3: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
7: ipudp00102@NONE: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1356 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ipip 172.31.39.81 peer 213.205.194.208
8: ipudp00101@NONE: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1356 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ipip 172.31.39.81 peer 85.255.236.215
13: ipudp00103@NONE: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1356 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ipip 172.31.39.81 peer 82.132.241.77

Output of dmesg:

[16499.813823] mptcp_alloc_mpcb: created mpcb with token 0x47566be5
[16499.816966] mptcp_add_sock: token 0x47566be5 pi 1, src_addr:172.31.39.81:35660 dst_addr:192.168.248.6:5001, cnt_subflows now 1
[16499.885992] mptcp_add_sock: token 0x47566be5 pi 2, src_addr:0.0.0.0:0 dst_addr:0.0.0.0:0, cnt_subflows now 2
[16499.890888] mptcp_init4_subsockets: token 0x47566be5 pi 2 src_addr:172.31.39.81:0 dst_addr:192.168.248.7:5001 ifidx: 2
[16499.895721] mptcp_add_sock: token 0x47566be5 pi 3, src_addr:0.0.0.0:0 dst_addr:0.0.0.0:0, cnt_subflows now 3
[16499.900062] mptcp_init4_subsockets: token 0x47566be5 pi 3 src_addr:172.31.39.81:0 dst_addr:192.168.248.5:5001 ifidx: 2
[16505.447707] mptcp_close: Close of meta_sk with tok 0x47566be5
[16506.318007] mptcp_del_sock: Removing subsock tok 0x47566be5 pi:3 state 7 is_meta? 0
[16506.321705] mptcp_del_sock: Removing subsock tok 0x47566be5 pi:2 state 7 is_meta? 0
[16506.357877] mptcp_del_sock: Removing subsock tok 0x47566be5 pi:1 state 7 is_meta? 0
[16506.361583] mptcp_sock_destruct destroying meta-sk token 0x47566be5

Host B
------

Linux-4.9.60

$ cat /proc/net/mptcp_fullmesh 
Index, Address-ID, Backup, IP-address
IPv4, next v4-index: 1
0, 1, 0, 192.168.248.7
1, 2, 0, 192.168.248.6
2, 3, 0, 192.168.248.5
IPv6, next v6-index: 0

$ ip link
1: lo: <LOOPBACK,NOMULTIPATH,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: eth0: <BROADCAST,MULTICAST,NOMULTIPATH,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT group default qlen 1000
    link/ether 52:54:00:b2:a1:d9 brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST,NOMULTIPATH,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT group default qlen 1000
    link/ether 52:54:00:c6:7d:5d brd ff:ff:ff:ff:ff:ff
4: eth2: <BROADCAST,MULTICAST,NOMULTIPATH,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT group default qlen 1000
    link/ether 52:54:00:76:e6:7d brd ff:ff:ff:ff:ff:ff
5: wwan2-0: <POINTOPOINT,MULTICAST,NOMULTIPATH,NOARP,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UNKNOWN mode DEFAULT group default qlen 1000
    link/none 
6: wwan2-1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether d2:51:66:8d:0d:7f brd ff:ff:ff:ff:ff:ff
7: wwan1-0: <POINTOPOINT,MULTICAST,NOMULTIPATH,NOARP,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UNKNOWN mode DEFAULT group default qlen 1000
    link/none 
8: wwan1-1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether d2:51:66:8d:0d:7f brd ff:ff:ff:ff:ff:ff
9: wwan0-0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether d2:51:66:8d:0d:7f brd ff:ff:ff:ff:ff:ff
10: wwan0-1: <BROADCAST,MULTICAST,NOMULTIPATH,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether d2:51:66:8d:0d:7f brd ff:ff:ff:ff:ff:ff
11: wlan0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DORMANT group default qlen 1000
    link/ether ac:fd:ce:c7:95:7f brd ff:ff:ff:ff:ff:ff
12: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN mode DEFAULT group default qlen 1
    link/ipip 0.0.0.0 brd 0.0.0.0
13: ipudp00102@NONE: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1356 qdisc htb state UNKNOWN mode DEFAULT group default qlen 1
    link/ipip 10.211.247.76 peer 35.178.63.250
14: ipudp00101@NONE: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1356 qdisc htb state UNKNOWN mode DEFAULT group default qlen 1
    link/ipip 10.27.40.246 peer 35.178.63.250
19: ipudp00103@NONE: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1356 qdisc htb state UNKNOWN mode DEFAULT group default qlen 1
    link/ipip 10.161.107.209 peer 35.178.63.250

Packet capture (host A)
-------------------------------
[mptcp_fou_no_join.zip](https://github.com/multipath-tcp/mptcp/files/1842211/mptcp_fou_no_join.zip)

**Comment by cpaasch** [comment]
```
[16499.813823] mptcp_alloc_mpcb: created mpcb with token 0x47566be5
[16499.816966] mptcp_add_sock: token 0x47566be5 pi 1, src_addr:172.31.39.81:35660 dst_addr:192.168.248.6:5001, cnt_subflows now 1
[16499.885992] mptcp_add_sock: token 0x47566be5 pi 2, src_addr:0.0.0.0:0 dst_addr:0.0.0.0:0, cnt_subflows now 2
[16499.890888] mptcp_init4_subsockets: token 0x47566be5 pi 2 src_addr:172.31.39.81:0 dst_addr:192.168.248.7:5001 ifidx: 2
[16499.895721] mptcp_add_sock: token 0x47566be5 pi 3, src_addr:0.0.0.0:0 dst_addr:0.0.0.0:0, cnt_subflows now 3
[16499.900062] mptcp_init4_subsockets: token 0x47566be5 pi 3 src_addr:172.31.39.81:0 dst_addr:192.168.248.5:5001 ifidx: 2
```

This indicates that Host A is indeed trying ti establish the subflows. I guess there is something in your FOU-configuration that is wrong. When you are on Host A, can you ping the IP-addresses ```192.168.248.5``` and ```192.168.248.7``` ?

**Comment by kpeletidis** [comment]
Good point. In this example 192.168.248.6 is used instead of 192.168.248.7. It can be seen that on Host A each address is pingable only via its associated tunnel, while on Host B either tunnel can be used:

**Host A**
```
# ip tun
ipudp00101: ip/ip remote 185.69.145.209 local 172.31.0.140 ttl 225
ipudp00102: ip/ip remote 213.205.194.18 local 172.31.0.140 ttl 225
tunl0: any/ip remote any local any ttl inherit nopmtudisc

# ping -I ipudp00101 192.168.248.5
ping: Warning: source address might be selected on device other than ipudp00101.
PING 192.168.248.5 (192.168.248.5) from 172.31.0.140 ipudp00101: 56(84) bytes of data.
64 bytes from 192.168.248.5: icmp_seq=1 ttl=64 time=22.5 ms
64 bytes from 192.168.248.5: icmp_seq=2 ttl=64 time=338 ms
^C
--- 192.168.248.5 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 22.553/180.648/338.744/158.096 ms

# ping -I ipudp00102 192.168.248.5
ping: Warning: source address might be selected on device other than ipudp00102.
PING 192.168.248.5 (192.168.248.5) from 172.31.0.140 ipudp00102: 56(84) bytes of data.
^C
--- 192.168.248.5 ping statistics ---
4 packets transmitted, 0 received, 100% packet loss, time 3060ms

# ping -I ipudp00102 192.168.248.6
ping: Warning: source address might be selected on device other than ipudp00102.
PING 192.168.248.6 (192.168.248.6) from 172.31.0.140 ipudp00102: 56(84) bytes of data.
64 bytes from 192.168.248.6: icmp_seq=1 ttl=64 time=43.4 ms
64 bytes from 192.168.248.6: icmp_seq=2 ttl=64 time=90.0 ms
^C
--- 192.168.248.6 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 43.458/66.761/90.065/23.304 ms

# ping -I ipudp00101 192.168.248.6
ping: Warning: source address might be selected on device other than ipudp00101.
PING 192.168.248.6 (192.168.248.6) from 172.31.0.140 ipudp00101: 56(84) bytes of data.
^C
--- 192.168.248.6 ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 2028ms

```
**Host B**
```
$ ip tun
ipudp00102: ip/ip  remote 35.178.69.188  local 10.205.228.205  ttl 225 
ipudp00101: ip/ip  remote 35.178.69.188  local 10.25.159.46  ttl 225 
tunl0: any/ip  remote any  local any  ttl inherit  nopmtudisc

$ ping -I ipudp00101 172.31.0.140
PING 172.31.0.140 (172.31.0.140) from 192.168.248.5 ipudp00101: 56(84) bytes of data.
64 bytes from 172.31.0.140: icmp_req=1 ttl=64 time=23.4 ms
64 bytes from 172.31.0.140: icmp_req=2 ttl=64 time=75.1 ms
64 bytes from 172.31.0.140: icmp_req=3 ttl=64 time=176 ms
^C
--- 172.31.0.140 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 1999ms
rtt min/avg/max/mdev = 23.497/91.792/176.764/63.672 ms

$ ping -I ipudp00102 172.31.0.140
PING 172.31.0.140 (172.31.0.140) from 192.168.248.5 ipudp00102: 56(84) bytes of data.
64 bytes from 172.31.0.140: icmp_req=1 ttl=64 time=49.6 ms
64 bytes from 172.31.0.140: icmp_req=2 ttl=64 time=81.8 ms
64 bytes from 172.31.0.140: icmp_req=3 ttl=64 time=122 ms
^C
--- 172.31.0.140 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2000ms
rtt min/avg/max/mdev = 49.617/84.539/122.147/29.671 ms

$ /sbin/ifconfig | grep -A1 ipudp  
ipudp00101 Link encap:IPIP Tunnel  HWaddr   
          inet addr:192.168.248.5  P-t-P:192.168.248.5  Mask:255.255.254.0
--
ipudp00102 Link encap:IPIP Tunnel  HWaddr   
          inet addr:192.168.248.6  P-t-P:192.168.248.6  Mask:255.255.254.0

```
Is something wrong with FOU then that interferes with MPTCP?

**Comment by pgreenland** [comment]
Hi Christoph et al,

I've picked up this issue from Kostas our end.

I believe I've got to the bottom of it and have a rough understanding of why things aren't working as expected.

Performing the same packet capture as Kostas I can see the first TCP connection established from the client 172.31.39.81 to a listening server on a tunnel of my choosing, 192.168.248.5 for example.

In that flow I see the address announcements and with MPTCP debugging enabled I see the debug messages on the client indicating that sub-flow creation has started, however I see no MPTCP traffic on the other tunnels.

What I do see however is TCP SYN packets from 172.31.39.81 destined to 192.168.248.6 and 192.168.248.7 (the tunnel end points) leaving the system on the wrong interface, in this case eth0, destined for the default gateway, which in turn responds with an ICMP - network unreachable.

There are ip rules in place in order to correctly route traffic back over the tunnels, for example

```
$ ip rule
0:	from all lookup local
1000:	from all to 192.168.248.5 lookup 201
1000:	from all to 192.168.248.6 lookup 202
1000:	from all to 192.168.248.7 lookup 203
32766:	from all lookup main
32767:	from all lookup default
```

However these at first appeared to be ignored by MPTCP.

I think I've found the problem in our case. Comparing our config to our much older but working 3.18 kernel, I noticed a new column "if-idx" in the fullmesh proc file "/proc/net/mptcp_fullmesh".

I also found "the smoking gun" https://github.com/multipath-tcp/mptcp/blob/d6cd22e29f5c9b7bf374e259bbb2d51be5cfaee5/net/mptcp/mptcp_ipv4.c#L325

Following the program flow around, `mptcp_init4_subsockets` calls, `kernel_connect` which via some indirection calls `tcp_v4_connect`, which in turn calls `ip_route_connect` passing `sk->sk_bound_dev_if` as the `oif` argument.

It would appear that MPTCP is only considering routes which use the output interface that owns the IP.

A quick bit of google foo returned the "ip route get" command. The correct route would be:

```
$ ip route get 192.168.248.6 from 172.31.39.81
192.168.248.6 from 172.31.39.81 dev ipudp00102 table 202 
    cache
```

Whereas when limited to a particular interface an incorrect route is selected:

```
$ ip route get 192.168.248.6 from 172.31.39.81 oif eth0
192.168.248.6 from 172.31.39.81  dev eth0
    cache
```

If I remove line 325 from mptcp_ipv4.c (linked above) routing is considered over all links and the sub-flows are established correctly.

The change was made in this commit https://github.com/multipath-tcp/mptcp/commit/9e763119cccc65ab01d29028489a3c3f08e17510 I can see why it was done in this case, any ideas on how our issue could be resolved in light of this change?

Thanks,

Phil

**Comment by cpaasch** [comment]
Hello Phil,

thanks a lot for this analysis! I see what is happening. It appears as if we should instead of picking the ifindex from the interface, we should be doing a route-lookup to see where the flow *would* get routed and then use this to bind the flow to the interface.

Do you want to work on a patch for this?


Thanks,
Christoph

**Comment by pgreenland** [comment]
Hi Christoph,

Thanks for the fast reply, I'm happy to look at this and provide a patch.

The original commit comment provided a good explanation of why the change had been added.

> mptcp: Check interface index in fullmesh pm

>This patch introduces support for checking and binding MPTCP subflows to
interface (as well as address, which is already done). The problem was
discovered on measurement nodes used in the MONROE EU-project. Each node
is equipped with multiple mifis of the web ui kind, that act as small
routers and hand out an IP over DHCP. The problem is that they all give
out the same IP, thus, there will be multiple interfaces with the same
address.

>The fullmesh find_address() function only compares address family and
address, and will not detect that there are multiple interfaces available.
We solve this by storing the interface index together with each source
address, and comparing the index as well in find_address(). Similar
functionality should probably be implemented in all path managers in due
time.

The current code to me makes sense in that case, ensuring that a route is chosen on the correct interface when multiple interfaces with the same IP are present.

If we were to perform the route lookup given the src and dest IPs, before selecting the interface to bind to we'd break the scenario above? - Correct me if I'm wrong, I'm a noob when it comes to the internals of the Linux network stack.

I was originally going to suggest adding a /proc/sys parameter to control the source interface binding. Although it sounds like you have a better idea.

Any chance you could provide a bit more info?

Thanks,

Phil

**Comment by cpaasch** [comment]
Hello @pgreenland  - you are right it would break the other scenario. This is tricky here. I think, the only way to handle this better would be by having a more dynamic path-manager (cfr., steered through netlink).

**Comment by lenormf** [comment]
Hi,

Has any progress been made on this, in the latest releases? @pgreenland are you still open to coming up with a patch for this issue?

**Comment by pgreenland** [comment]
Hey @lenormf,

I must admit that I haven't looked at it past the point me and @cpaasch had got to above.

I went with the sysctl flag in our kernel build to disable the feature that was causing me problems.

Didn't seem like the best solution for submission as a patch though and was only meant to be temporary....but....as these things tend to go I moved onto the next fire.

If I do come up with something better I'll be submitting it back to the community for sure!

If you're interested in my "workaround" I'm happy to share.

Thanks,

Phil
